AlphaFold documentation (#736)
NREL/HPC: docs/Documentation/Applications/alphafold.md

# AlphaFold

[AlphaFold](https://github.com/google-deepmind/alphafold3/tree/main) is an open-source inference pipeline developed by Google DeepMind to predict three-dimensional protein structures from input biological sequence data. The default model from the third major release of AlphaFold (AlphaFold3) is currently only supported on Kestrel.

## Using the AlphaFold module

AlphaFold runs as a [containerized](../Development/Containers/index.md) module on Kestrel and can leverage GPU resources. As such, **you must be on a GPU node to load the module**: `ml alphafold`. Once the module is loaded, run `run_alphafold --help` for the pipeline's help page.

!!! Note

>     You need to be part of the `esp-alphafold3` group on Kestrel to access the model and its reference databases in `/kfs2/shared-projects/alphafold3`.

Reviewer: ```suggestion
    You need to be part of the `esp-alphafold3` group on Kestrel to access the model and its reference databases in `/kfs2/shared-projects/alphafold3`. To be added to the group, contact [HPC-Help@nrel.gov](mailto:HPC-Help@nrel.gov).
```

AlphaFold requires input data to be formatted in a [special JSON style](https://github.com/google-deepmind/alphafold3/blob/main/docs/input.md). Please see the link for the full breadth of available JSON elements, noting that anything specific to "AlphaFold Server" does not apply to the module on Kestrel. A simple input JSON file (`fold_input.json`) is provided below as an example. Refer to the next section for instructions on how to run AlphaFold on JSON files like this one.

??? example "Input JSON format example: `fold_input.json`"
    This is a simple example input JSON file for an AlphaFold run:
    ```json
    {
    "name": "2PV7",
    "sequences": [
        {
        "protein": {
            "id": ["A", "B"],
            "sequence": "GMRESYANENQFGFKTINSDIHKIVIVGGYGKLGGLFARYLRASGYPISILDREDWAVAESILANADVVIVSVPINLTLETIERLKPYLTENMLLADLTSVKREPLAKMLEVHTGAVLGLHPMFGADIASMAKQVVVRCDGRFPERYEWLLEQIQIWGAKIYQTNATEHDHNMTYIQALRHFSTFANGLHLSKQPINLANLLALSSPIYRLELAMIGRLFAQDAELYADIIMDKSENLAVIETLKQTYDEALTFFENNDRQGFIDAFHKVRDWFGDYSEQFLKESRQLLQQANDLKQG"
        }
        }
    ],
    "modelSeeds": [1],
    "dialect": "alphafold3",
    "version": 1
    }
    ```

## Running AlphaFold on Kestrel

> At the time of writing, AlphaFold3 is generally limited to running its inference pipeline on a single JSON on a single GPU device at a time. In other words, multi-GPU and/or multi-node jobs are not inherently supported by the software. As such, running the model sequentially on multiple JSONs requires the least amount of effort on the user's part. However, depending on the number of input JSONs, this limited throughput can lead to prohibitively long jobs in terms of research productivity. For cases in which you need to run AlphaFold on many JSON files representing as many individual sequences, we can instead leverage the "embarrassing parallelism" that Slurm provides with [job arrays](../Slurm/job_arrays.md). 

Reviewer: ```suggestion
As of February 2025, AlphaFold3 is generally limited to running its inference pipeline on a single JSON on a single GPU device at a time. In other words, multi-GPU and/or multi-node jobs are not inherently supported by the software. As such, running the model sequentially on multiple JSONs requires the least amount of effort on the user's part. However, depending on the number of input JSONs, this limited throughput can lead to prohibitively long jobs in terms of research productivity. For cases in which you need to run AlphaFold on many JSON files representing as many individual sequences, we can instead leverage the "embarrassing parallelism" that Slurm provides with [job arrays](../Slurm/job_arrays.md). 
```

Example submission scripts representing both strategies are given in the next sections.

!!! Note
    As for every GPU job, please ensure that you submit your AlphaFold jobs from one of the [GPU login nodes](../Systems/Kestrel/index.md).

### Low throughput: Sequentially loop through JSON files

If throughput is not a concern, running AlphaFold on several JSON files is straightforward. Simply load the AlphaFold module, save your JSON files into a folder that is found by the `--input_dir` option, and define an output directory with `--output_dir` in the `run_alphafold` command: 

??? example "Sequentially loop through input JSON files in a folder: `af3_sequential.sh`"
    ```slurm
    #!/bin/bash
    #SBATCH -A hpcapps # Replace with your HPC account

>     #SBATCH -p gpu-h100

Reviewer: ```suggestion
```

    #SBATCH --mem=80G # Note this is for system (i.e., CPU) RAM
    #SBATCH -o %x-%j.log
    
    # Load Alphafold3 module
    ml alphafold
    
    # Run Alphafold3 inference pipeline to predict 3d structure from sequence
    # Note: the "af_input" folder contains "fold_input.json"
    run_alphafold --input_dir af_input --output_dir af_output
    ```

### High throughput: "Embarrassing parallelism" using job arrays

If you need to run AlphaFold on many (e.g., dozens to hundreds) of JSON files, it is worth setting up a job array to run the model on each input in separate jobs from the same submission script. This is a form of "embarrassing parallelism," i.e., a form of parallel computing in which each task is independent of any other and therefore does not require communication between nodes. Thinking of each input JSON as its own "array task" allows us to submit separate jobs that can run simultaneously, which can significantly increase throughput when compared to the `af3_sequential.sh` example above:

??? example "Submit separate jobs for each input JSON file: `af3_job_array.sh`"
    ```slurm
    #!/bin/bash
    #SBATCH -A hpcapps # Replace with your HPC account

>     #SBATCH -p gpu-h100

Reviewer: ```suggestion
```
